dm3-names/contracts/l2/Dm3NameRegistrar.sol

```diff
@@ -0,0 +1,192 @@
+
+    // Constant for reverse node address
+    //Taken from ENS ReverseRegistrar contract
+    //https://github.com/ensdomains/ens-contracts/blob/21736916300b26cb8ea1802dbf6c9ff054adaeab/contracts/reverseRegistrar/ReverseRegistrar.sol#L12
+    bytes32 private constant ADDR_REVERSE_NODE =
+        0x91d1777781884d03a6757a803996e38de2a42967fb37eeaca72729271025a9e2;
+
+    //Node of the L1 domain. in case of OP name that would be namehash(op.dm3.eth)
+    //Is not immutable so it can be retrieved from the storage using CCIP
+    bytes32 public PARENT_NODE;
+
+    // Mapping to store the owner of each node
+    mapping(bytes32 => address) public owner;
+
+    // Mapping to store the reverse record of each node
+    mapping(bytes32 => string) public reverse;
+
+    // Mapping to store text records for each node
+    mapping(uint64 => mapping(bytes32 => mapping(string => string)))
+        public texts;
+
+    // Event emitted when a name is registered
+    event NameRegistered(address indexed addr, string indexed name);
+
+    // Event emitted when a name is removed
+    event NameRemoved(address indexed addr, string indexed name);
+
+    function initialize(bytes32 _parentNode) public initializer {
+        PARENT_NODE = _parentNode;
+    }
+    function isAuthorised(bytes32 node) internal view override returns (bool) {
+        return owner[node] == msg.sender;
+    }
+
+    /// @notice Register a name in the ENS system
+    /// @param _name The name to register
+    function register(string calldata _name) external {
+        string memory oldName = reverse[makeReverseNode(msg.sender)];
+        //Delete name
+        if (bytes(_name).length == 0) {
+            //clear text records
+            clearRecords(makeLabelNode(oldName));
+            // Clear name if the new name is empty
+            delete owner[makeLabelNode(oldName)];
+            delete reverse[makeReverseNode(msg.sender)];
+            emit NameRemoved(msg.sender, oldName);
+            return;
+        }
+        //Check if the name is already registered. If not, revert
+        require(
+            owner[makeLabelNode(_name)] == address(0),
+            'Name already registered'
+        );
+        //Select new name
+        if (bytes(oldName).length > 0) {
+            //clear text records
+            clearRecords(makeLabelNode(oldName));
+            // Clear old name if it exists
+            delete owner[makeLabelNode(oldName)];
+            emit NameRemoved(msg.sender, oldName);
+        }
+        _mint(_name, msg.sender);
+    }
+
+    /// @notice Allows the contract owner to mint a new name
+    /// @dev This function can only be called by the contract owner
+    /// @param _name The name to be minted
+    /// @param _owner The address that will own the minted name
+    function ownerRegister(
+        string calldata _name,
+        address _owner
+    ) external onlyOwner {
+        //Even the owner can't mint a name that is already registered
+        require(
```

Comment:
because we do not check if a name is already assigned to `_owner`, this function allows the registration of multiple names to one address. This seems a bad idea (e.g. the reverse record would most likely point to the last name registered, and `owner` could register a name even if the `_owner` already has a name). 

Also, giving `_owner` and `owner` completely different meanings might lead to misunderstandings down the line. How about `_user` instead of `_owner`?

Reply:
I know that the ambiguity between the two FN selectors is not ideal, however, the ENSRegistry interface specifies a method owner(bytes32) that returns the address of that owner. 
I'd like to keep that function so our registrar is compliant with ENS. A solution for that might be to rename the function provided by the Ownable contract. Not sure if that would be the cleanest solution either though.

Reply:
I see your point. Let's keep both `owner()` and `owner(bytes32)` and document this decision.